File: leetcode_py/Ugly Number II.py
class Solution(object):
    def nthUglyNumber(self, n):
        """
        :type n: int
        :rtype: int
        """
        res = [1]
        idx2, idx3, idx5 = 0, 0, 0
        for i in range(n - 1):
            res.append(min(res[idx2] * 2, res[idx3] * 3, res[idx5] * 5))
            if res[-1] == res[idx2] * 2:
                idx2 += 1
            if res[-1] == res[idx3] * 3:
                idx3 += 1
            if res[-1] == res[idx5] * 5:
                idx5 += 1
        return res[-1]


if __name__ == "__main__":
    a = Solution()
    print a.nthUglyNumber(7)

File: leetcode_js/Ugly Number II.js
/**
 * @param {number} n
 * @return {number}
 */
var nthUglyNumber = function(n) {
    var res = [1], cur = idx2 = idx3 = idx5 = 0;
    while(cur < n-1){
        res[++cur] = Math.min(res[idx2]*2, res[idx3]*3, res[idx5]*5);
        if(res[idx2]*2 === res[cur]) idx2++;
        if(res[idx3]*3 === res[cur]) idx3++;
        if(res[idx5]*5 === res[cur]) idx5++;
    }
    return res[n-1];
};
document.writeln(nthUglyNumber(7));